File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/textures/MultiTexture.java
package org.saar.lwjgl.opengl.textures;

import java.util.function.Consumer;

public class MultiTexture implements ReadOnlyTexture {

    private final ReadOnlyTexture blendMap; // Blend Map Texture
    private final ReadOnlyTexture dTexture; // Default   Texture
    private final ReadOnlyTexture rTexture; // Red       Texture
    private final ReadOnlyTexture gTexture; // Green     Texture
    private final ReadOnlyTexture bTexture; // Blue      Texture

    public MultiTexture(ReadOnlyTexture blendMap, ReadOnlyTexture dTexture, ReadOnlyTexture rTexture,
                        ReadOnlyTexture gTexture, ReadOnlyTexture bTexture) {
        this.blendMap = blendMap;
        this.dTexture = dTexture;
        this.rTexture = rTexture;
        this.gTexture = gTexture;
        this.bTexture = bTexture;
    }

    @Override
    public void bind(int unit) {
        this.blendMap.bind(unit);
        this.dTexture.bind(unit + 1);
        this.rTexture.bind(unit + 2);
        this.gTexture.bind(unit + 3);
        this.bTexture.bind(unit + 4);
    }

    @Override
    public void bind() {
        this.bind(0);
    }

    @Override
    public void unbind() {
        this.forEach(ReadOnlyTexture::unbind);
    }

    @Override
    public void delete() {
        this.forEach(ReadOnlyTexture::delete);
    }

    private void forEach(Consumer<ReadOnlyTexture> consumer) {
        consumer.accept(this.blendMap);
        consumer.accept(this.dTexture);
        consumer.accept(this.rTexture);
        consumer.accept(this.gTexture);
        consumer.accept(this.bTexture);
    }
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/texture/MultiTexture.kt
package org.saar.lwjgl.opengl.texture

class MultiTexture(
    private val blendMap: ReadOnlyTexture, // Blend Map Texture
    private val dTexture: ReadOnlyTexture, // Default   Texture
    private val rTexture: ReadOnlyTexture, // Red       Texture
    private val gTexture: ReadOnlyTexture, // Green     Texture
    private val bTexture: ReadOnlyTexture, // Blue      Texture
) : ReadOnlyTexture {

    override fun bind(unit: Int) = forEach { texture, index -> texture.bind(unit + index) }

    override fun bind() = bind(0)

    override fun unbind() = forEach { obj, _ -> obj.unbind() }

    override fun delete() = forEach { obj, _ -> obj.delete() }

    private inline fun forEach(consumer: (ReadOnlyTexture, Int) -> Unit) {
        consumer(this.blendMap, 0)
        consumer(this.dTexture, 1)
        consumer(this.rTexture, 2)
        consumer(this.gTexture, 3)
        consumer(this.bTexture, 4)
    }
}